Use a timer to purged old matched rules

b-weinstein opened the pull request:
#### 5a1eb5d9c5cec2c7f696dc953963f56f246dc699
<pre>
Use a timer to purged old matched rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=265950">https://bugs.webkit.org/show_bug.cgi?id=265950</a>
<a href="https://rdar.apple.com/118940129">rdar://118940129</a>

Reviewed by Timothy Hatcher.

When a content rule list notification is saved as a matched rule by an extension, start a repeating 5 minute timer to purge
any rules that are older than 5 minutes.

If after the purge there aren&apos;t any remaining saved rules, stop the timer.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::handleContentRuleListNotificationForTab): This now returns a bool representing if the
notification was saved as a matched rule.
(WebKit::WebExtensionContext::purgeMatchedRulesFromBefore): Iterate over the rules and remove any that were from before the given
start time.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification): Start a timer if the action was saved and there wasn&apos;t a timer
started already.
(WebKit::WebExtensionController::purgeOldMatchedRules): Iterate over all of the extensions and call purgeMatchedRulesFromBefore.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules): Remove an extra newline.

Canonical link: <a href="https://commits.webkit.org/271629@main">https://commits.webkit.org/271629@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6d8ffa1db6859a0defad4f4405e42385eb89b9db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29075 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7747 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30419 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31681 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9873 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5055 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29347 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6404 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5552 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33018 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7320 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6934 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6180 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -52,6 +52,8 @@
 #import <wtf/NeverDestroyed.h>
 #import <wtf/text/WTFString.h>
 
+static constexpr double purgeMatchedRulesIntervalInMinutes = 5;
```
`static constexpr Seconds purgeMatchedRulesInterval = 5_min;`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -307,12 +311,32 @@
 
+void WebExtensionController::purgeOldMatchedRules()
+{
+    WallTime earliestDateToKeep = WallTime::now() - Seconds::fromMinutes(purgeMatchedRulesIntervalInMinutes);
```
`WallTime earliestDateToKeep = WallTime::now() - purgeMatchedRulesInterval;`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -307,12 +311,32 @@
     }
+
+    if (!savedMatchedRule)
+        return;
```
```
if (!savedMatchedRule || m_purgeOldMatchedRulesTimer)
    return;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -307,12 +311,32 @@
+    if (!savedMatchedRule)
+        return;
+
+    if (!m_purgeOldMatchedRulesTimer) {
```
Then you can get rid of this branch.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -307,12 +311,32 @@
+{
+    WallTime earliestDateToKeep = WallTime::now() - Seconds::fromMinutes(purgeMatchedRulesIntervalInMinutes);
+
+    bool stillHaveRules = false;
+    for (auto& extensionContext : m_extensionContexts)
+        stillHaveRules |= extensionContext->purgeMatchedRulesFromBefore(earliestDateToKeep);
+
+    if (!stillHaveRules)
+        m_purgeOldMatchedRulesTimer = nullptr;
```
I don't think you need the `bool` return and `stillHaveRules`. This can be:

```
if (m_matchedRules.isEmpty())
    m_purgeOldMatchedRulesTimer = nullptr;
```

b-weinstein replied:
This is in the controller so it applies across all extensions - `m_matchedRules` is per extension.

b-weinstein replied:
So that means we need to keep track of if _any_ extensions still have rules - not just one of them.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3013,20 +3013,35 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+            filteredMatchedRules.append(matchedRule);
+    }
+
+    m_matchedRules = filteredMatchedRules;
```
`std::exchange(m_matchedRules, filteredMatchedRules);`

b-weinstein replied:
Adopted `m_matchedRules = WTFMove(filteredMatchedRules);`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3013,20 +3013,35 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    return true;
+}
+
+bool WebExtensionContext::purgeMatchedRulesFromBefore(const WallTime& startTime)
```
Ditto. I don't think the `bool` return is needed.

b-weinstein replied:
I responded to that comment. I think it is necessary due to the fact one of these is on the controller the other is on the context.

b-weinstein committed:
Use a timer to purged old matched rules
https://bugs.webkit.org/show_bug.cgi?id=265950
rdar://118940129

Reviewed by Timothy Hatcher.

When a content rule list notification is saved as a matched rule by an extension, start a repeating 5 minute timer to purge
any rules that are older than 5 minutes.

If after the purge there aren't any remaining saved rules, stop the timer.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::handleContentRuleListNotificationForTab): This now returns a bool representing if the
notification was saved as a matched rule.
(WebKit::WebExtensionContext::purgeMatchedRulesFromBefore): Iterate over the rules and remove any that were from before the given
start time.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification): Start a timer if the action was saved and there wasn't a timer
started already.
(WebKit::WebExtensionController::purgeOldMatchedRules): Iterate over all of the extensions and call purgeMatchedRulesFromBefore.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules): Remove an extra newline.

Canonical link: https://commits.webkit.org/271629@main
